### PR TITLE
SQL: Support for escape sequences

### LINKED ISSUE #31883: SQL: Extend grammar to support JDBC/ODBC escape sequences
[Escape sequences](https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/odbc-escape-sequences?view=sql-server-2017) are typically used by tools to invoke functions or pass in date/time in a common format.

Which is pretty important for interoperability purposes.

### PR DESCRIPTION
**Pull request by costin**
Enhance grammar to allow JDBC/ODBC escape sequences, namely
- date, time and timestamp {d ''}, {t ''} and {ts ''}
- guid {guid ''}
- LIKE escape {escape ''}
- scalar function {fn }

Fix #31883

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/EscapedFunctionsTests.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                "line 1:2: Invalid timestamp received; Invalid format: \"2012-01-01T10:01:02.3456\" is malformed at \"T10:01:02.3456\"",
+                ex.getMessage());
+    }
+
+    public void testGUID() {
+        Literal l = guidLiteral("12345678-90ab-cdef-0123-456789abcdef");
+        assertThat(l.dataType(), is(DataType.KEYWORD));
+    }
+
+    public void testGUIDValidation() {
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/EscapedFunctionsTests.java]: Worth adding some more tests here? Like uppercase/lowercase combined letters and other letters from the alphabet?

**Review by astefan — APPROVED** [review]

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/EscapedFunctionsTests.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/EscapedFunctionsTests.java]: The upper/lower case mix is addressed in the following test. Nevertheless I've added a couple of tests that check the different validations of GUID.

**Comment by costin** [comment]
run gradle build tests

File: x-pack/plugin/sql/src/main/antlr/SqlBase.g4 [hunk 2]
```diff
@@ -197,25 +207,55 @@ valueExpression
...
+castTemplate
+    : CAST '(' expression AS dataType ')'
+    ;
+    
+extractExpression
+    : extractTemplate
+    | FUNCTION_ESC extractTemplate ESC_END
+    ;
+    
+extractTemplate
```
Comment (nik9000) [inline x-pack/plugin/sql/src/main/antlr/SqlBase.g4]: Why call these `xxTemplate`?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java [hunk 3]
```diff
@@ -222,17 +235,18 @@ public LikePattern visitPattern(PatternContext ctx) {
         }
 
         char escape = 0;
-        String escapeString = string(ctx.escape);
+        PatternEscapeContext escapeCtx = ctx.patternEscape();
+        String escapeString = string(escapeCtx != null ? escapeCtx.escape : null);
```
Comment (nik9000) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java]: `escapeCtx == null ? null : string(escapeCtx.escape);` feels a little cleaner to me. No need to pitch `null` through `string` and make the reader wonder what `string` does with null.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/SqlBaseBaseListener.java [hunk 4]
```diff
@@ -1,8 +1,3 @@
-/*
```
Comment (nik9000) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/SqlBaseBaseListener.java]: ++

**Review by nik9000 — APPROVED** [review]
LGTM.

I left a question about naming but :man_shrugging:, the name is fine, just not one I would have used.

**Review by bpintea — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/antlr/SqlBase.g4]: Out of names to use (which need to be unique between rules _and_ callbacks).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java]: +1

**Pull request merged**